Add ability to functorch tests via run_test.py

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82013
* __->__ #82012

This PR:
- adds the ability to run functorch tests via run_test.py
- changes the functorch shards in PyTorch CI to invoke functorch tests
via run_test.py

The main motivation for this is so that functorch tests hook into the
standard PyTorch test infrastructure.

Questions for reviewers:
- the functorch tests are located outside of the pytorch/test folder
(they're in the pytorch/functorch/test folder). Is this OK? (run_test.py
works locally for me).

Test Plan:

- checked that `python run_test.py --functorch` ran functorch tests
locally
- Local mock test: added `{"test_compilation_for_dynamic_shape
(__main__.TestCompileCache)":
["https://github.com/pytorch/pytorch/issues/82016", ["linux"]]}` to .pytorch-disabled-tests.json, ran functorch tests, verified that the test was skipped.
- Wait for CI

zou3519 committed:
Add ability to functorch tests via run_test.py

This PR:
- adds the ability to run functorch tests via run_test.py
- changes the functorch shards in PyTorch CI to invoke functorch tests
via run_test.py

The main motivation for this is so that functorch tests hook into the
standard PyTorch test infrastructure.

Questions for reviewers:
- the functorch tests are located outside of the pytorch/test folder
(they're in the pytorch/functorch/test folder). Is this OK? (run_test.py
works locally for me).

Test Plan:
- I tested run_test.py locally
- Wait for CI

[ghstack-poisoned]

janeyx99 commented on test/run_test.py:
```diff
@@ -317,6 +317,19 @@ def skip_test_p(name: str) -> bool:
 DISTRIBUTED_TESTS = [test for test in TESTS if test.startswith("distributed")]
 
+
+def discover_functorch_tests():
```
Some of our linters assume that all tests are in the test folder (see our test ownership lint). Would it be bad to move functorch tests to be test/functorch?

zou3519 replied:
We can't move all functorch tests from functorch/test to the test/ folder yet. because we are still syncing functorch to the pytorch/functorch repository.

We're planning on stopping that sync in around a month, at which point we can actually move functorch tests from functorch/test to the test folder if that would make things better. Is this OK?

Also, it sounds like I should add some test ownership comments to the functorch tests?

zou3519 replied:
in the latest update I added some test owner comments to the test files

janeyx99 replied:
can we have an issue to track moving functorch tests under test/ so we don’t forget about it?

zou3519 replied:
https://github.com/pytorch/pytorch/issues/82109

zou3519 committed:
Update on "Add ability to functorch tests via run_test.py"


This PR:
- adds the ability to run functorch tests via run_test.py
- changes the functorch shards in PyTorch CI to invoke functorch tests
via run_test.py

The main motivation for this is so that functorch tests hook into the
standard PyTorch test infrastructure.

Questions for reviewers:
- the functorch tests are located outside of the pytorch/test folder
(they're in the pytorch/functorch/test folder). Is this OK? (run_test.py
works locally for me).

Test Plan:

- checked that `python run_test.py --functorch` ran functorch tests
locally
- Local mock test: added `{"test_compilation_for_dynamic_shape
(__main__.TestCompileCache)":
["https://github.com/pytorch/pytorch/issues/82016", ["linux"]]}` to .pytorch-disabled-tests.json, ran functorch tests, verified that the test was skipped.
- Wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Add ability to functorch tests via run_test.py"


This PR:
- adds the ability to run functorch tests via run_test.py
- changes the functorch shards in PyTorch CI to invoke functorch tests
via run_test.py

The main motivation for this is so that functorch tests hook into the
standard PyTorch test infrastructure.

Questions for reviewers:
- the functorch tests are located outside of the pytorch/test folder
(they're in the pytorch/functorch/test folder). Is this OK? (run_test.py
works locally for me).

Test Plan:

- checked that `python run_test.py --functorch` ran functorch tests
locally
- Local mock test: added `{"test_compilation_for_dynamic_shape
(__main__.TestCompileCache)":
["https://github.com/pytorch/pytorch/issues/82016", ["linux"]]}` to .pytorch-disabled-tests.json, ran functorch tests, verified that the test was skipped.
- Wait for CI

[ghstack-poisoned]

janeyx99 commented on .jenkins/pytorch/common_utils.sh:
```diff
@@ -145,6 +145,10 @@ function install_functorch() {
   popd
 }
 
+function test_functorch() {
```
why move the test_functorch function to common_utils from test.sh?

zou3519 replied:
I want to use this in https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/macos-test.sh when I add macos tests and this file imports common_utils.sh

janeyx99 reviewed (approved):
overall good, could you also verify the following:
1. sharding works: python run_test.py —functorch -shard 1 2 should be disjoint from using -shard 2 2 and together should run all tests 
2. test reports are actually uploaded—you can go to the hud pr page (first link in the dr. ci comment) and download the test reports for the config that runs test_functorch + make sure you see all relevant test reports

zou3519 commented:
Thanks for the review and list of things to verify.

> 1. sharding works: python run_test.py —functorch -shard 1 2 should be disjoint from using -shard 2 2 and together should run all tests

I tested that sharding works

> 2. test reports are actually uploaded—you can go to the hud pr page (first link in the dr. ci comment) and download the test reports for the config that runs test_functorch + make sure you see all relevant test reports

The test reports exist, but the file path is different (and probably depends on the machine). We probably want to post-process them somehow

<img width="1250" alt="image" src="https://user-images.githubusercontent.com/5652049/180797879-899416f3-1994-43e8-b1a0-513b7e8021be.png">
